README.md
# SSKTS-WAITER

[![CircleCI](https://circleci.com/gh/motionpicture/sskts-waiter.svg?style=svg&circle-token=1e1bf1d7d27e8c8ba45e8a44da2a5640bcbee147)](https://circleci.com/gh/motionpicture/sskts-waiter)

## Table of contents

* [Background](#background)
* [Usage](#usage)
* [Jsdoc](#jsdoc)
* [License](#license)

## Background
see [waiter repository](https://github.com/motionpicture/waiter)

## Usage

### インフラ
#### web server
node.js application  
- nginx on [GCP AppEngine](https://cloud.google.com/appengine/?hl=ja)

#### DB
- Redis Cache

### Environment variables

| Name                     | Required | Value    | Purpose                                   |
| ------------------------ | -------- | -------- | ----------------------------------------- |
| `DEBUG`                  | false    | waiter:* | Debug                                     |
| `NODE_ENV`               | true     |          | 許可証暗号化の秘密鍵                       |
| `WAITER_PASSPORT_ISSUER` | true     |          | 許可証発行者識別子(通常発行APIのドメインを指定) |
| `WAITER_RULES`           | true     |          | 発行規則リスト                               |
| `WAITER_SECRET`          | true     |          | 許可証暗号化の秘密鍵                       |
| `REDIS_HOST`             | true     |          | Redis Cache接続ホスト                        |
| `REDIS_PORT`             | true     |          | Redis Cache接続ポート                        |
| `REDIS_KEY`              | true     |          | Redis Cache接続キー                         |

## Jsdoc

`npm run doc`でjsdocを作成できます。./docに出力されます。

## License

UNLICNSED


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/).

## Unreleased

### Added

### Changed

### Deprecated

### Removed

### Fixed

### Security

## v2.0.1 - 2018-06-13

### Security

- 一時的にnspの例外を追加。

## v2.0.0 - 2017-12-02

### Added

- 本家WAITERを引き継いでv2.0.0としてリリース。


src/app/middlewares/errorHandler.ts
/**
 * error handler
 * エラーハンドラーミドルウェア
 * @module middlewares.errorHandler
 */

import * as waiter from '@motionpicture/waiter-domain';
import { NextFunction, Request, Response } from 'express';
import { BAD_REQUEST, FORBIDDEN, INTERNAL_SERVER_ERROR, NOT_FOUND, SERVICE_UNAVAILABLE, TOO_MANY_REQUESTS } from 'http-status';

import { APIError } from '../error/api';

export default (err: any, __: Request, res: Response, next: NextFunction) => {
    if (res.headersSent) {
        next(err);

        return;
    }

    let apiError: APIError;
    if (err instanceof APIError) {
        apiError = err;
    } else {
        if (err instanceof waiter.factory.errors.Waiter) {
            switch (true) {
                // 401
                // case (err instanceof waiter.factory.errors.Unauthorized):
                //     apiError = new APIError(UNAUTHORIZED, [err]);
                //     break;

                // 403
                case (err instanceof waiter.factory.errors.Forbidden):
                    apiError = new APIError(FORBIDDEN, [err]);
                    break;

                // 404
                case (err instanceof waiter.factory.errors.NotFound):
                    apiError = new APIError(NOT_FOUND, [err]);
                    break;

                // 409
                // case (err instanceof waiter.factory.errors.AlreadyInUse):
                //     apiError = new APIError(CONFLICT, [err]);
                //     break;

                // 409
                case (err instanceof waiter.factory.errors.RateLimitExceeded):
                    apiError = new APIError(TOO_MANY_REQUESTS, [err]);
                    break;

                // 503
                case (err instanceof waiter.factory.errors.ServiceUnavailable):
                    apiError = new APIError(SERVICE_UNAVAILABLE, [err]);
                    break;

                // 400
                default:
                    apiError = new APIError(BAD_REQUEST, [err]);
                    break;
            }
        } else {
            // 500
            apiError = new APIError(INTERNAL_SERVER_ERROR, [new waiter.factory.errors.Waiter(<any>'InternalServerError', err.message)]);
        }
    }

    res.status(apiError.code).json({
        error: apiError.toObject()
    });
};


dst/app/middlewares/errorHandler.js
"use strict";
/**
 * error handler
 * エラーハンドラーミドルウェア
 * @module middlewares.errorHandler
 */
Object.defineProperty(exports, "__esModule", { value: true });
const waiter = require("@motionpicture/waiter-domain");
const http_status_1 = require("http-status");
const api_1 = require("../error/api");
exports.default = (err, __, res, next) => {
    if (res.headersSent) {
        next(err);
        return;
    }
    let apiError;
    if (err instanceof api_1.APIError) {
        apiError = err;
    }
    else {
        if (err instanceof waiter.factory.errors.Waiter) {
            switch (true) {
                // 401
                // case (err instanceof waiter.factory.errors.Unauthorized):
                //     apiError = new APIError(UNAUTHORIZED, [err]);
                //     break;
                // 403
                case (err instanceof waiter.factory.errors.Forbidden):
                    apiError = new api_1.APIError(http_status_1.FORBIDDEN, [err]);
                    break;
                // 404
                case (err instanceof waiter.factory.errors.NotFound):
                    apiError = new api_1.APIError(http_status_1.NOT_FOUND, [err]);
                    break;
                // 409
                // case (err instanceof waiter.factory.errors.AlreadyInUse):
                //     apiError = new APIError(CONFLICT, [err]);
                //     break;
                // 409
                case (err instanceof waiter.factory.errors.RateLimitExceeded):
                    apiError = new api_1.APIError(http_status_1.TOO_MANY_REQUESTS, [err]);
                    break;
                // 503
                case (err instanceof waiter.factory.errors.ServiceUnavailable):
                    apiError = new api_1.APIError(http_status_1.SERVICE_UNAVAILABLE, [err]);
                    break;
                // 400
                default:
                    apiError = new api_1.APIError(http_status_1.BAD_REQUEST, [err]);
                    break;
            }
        }
        else {
            // 500
            apiError = new api_1.APIError(http_status_1.INTERNAL_SERVER_ERROR, [new waiter.factory.errors.Waiter('InternalServerError', err.message)]);
        }
    }
    res.status(apiError.code).json({
        error: apiError.toObject()
    });
};
